### PR TITLE
enabled T_LIMIT_REQ macro for limit req module enhancement

### PR DESCRIPTION
**Pull request by hongxiaolong**
base branch from nginx master(ahead of stable-1.12) with commit as follows:

commit 7b9bc5d4d1c061095a2391dbe32cf4d205c031bc
Author: Maxim Dounin <mdounin@mdounin.ru>
Date:   Wed May 30 15:40:34 2018 +0300

### CODE REVIEW

**Comment by hongxiaolong** [comment]
* Tested-by: HONGXIAOLONG                                                                                                                                                                                   * Date: Thu Jul 26 21:33:56 CST 2018
* hostname: CentOS-6.9
* pwd: /root/Github/hongxiaolong/tengine
* cases: tests/nginx-tests/nginx-tests/limit_req.t
* git log:

```
commit 6bd30db85707ad7adf000532de7a4c34d6f0c986
Author: HONGXIAOLONG <imhongxiaolong@gmail.com>
Date:   Fri Jul 27 20:25:20 2018 +0800

    enabled T_LIMIT_REQ macro for limit req module enhancement
```

* case result:

```
$ MOD_DYCONF_PATH=/root/Github/hongxiaolong/tengine TEST_NGINX_BINARY=/root/Github/hongxiaolong/tengine/objs/nginx prove -v -I /root/Github/hongxiaolong/tengine/tests/nginx-tests/nginx-tests/lib tests/ngi
nx-tests/nginx-tests/limit_req.t
tests/nginx-tests/nginx-tests/limit_req.t ..
1..6
qok 1 - request
ok 2 - request rejected
ok 3 - delayed big request not truncated
ok 4 - rejects not counted
ok 5 - negative excess
ok 6 - no alerts
ok
All tests successful.
Files=1, Tests=6,  1 wallclock secs ( 0.01 usr  0.00 sys +  0.07 cusr  0.02 csys =  0.10 CPU)
Result: PASS
```

* Report-generated-by: http://wxc.oss.aliyuncs.com/dtest.sh

File: src/http/ngx_http_request.c [hunk 1]
```diff
@@ -2240,6 +2240,11 @@ ngx_http_request_handler(ngx_event_t *ev)
     ngx_log_debug2(NGX_LOG_DEBUG_HTTP, c->log, 0,
                    "http run request: \"%V?%V\"", &r->uri, &r->args);
 
+    if (ev->delayed && ev->timedout) {
+        ev->delayed = 0;
+        ev->timedout = 0;
+    }
+
```
Comment (hongxiaolong) [inline src/http/ngx_http_request.c]: related PR: http://hg.nginx.org/nginx/rev/903fb1ddc07f

**Comment by chobits** [comment]
hi @hongxiaolong 

Should add a diff output between this module and nginx-1.15.1, which make reviewing easy.

**Comment by hongxiaolong** [comment]
@chobits diff between my PR and nginx master(_d5c43bf51c9c62d26c4c2c09a4a337d8b8b1c060as_) as follows:

```
--- tengine/src/http/modules/ngx_http_limit_req_module.c	2018-07-30 14:56:15.000000000 +0800
+++ nginx/src/http/modules/ngx_http_limit_req_module.c	2018-07-30 15:02:08.000000000 +0800
@@ -9,12 +9,6 @@
 #include <ngx_core.h>
 #include <ngx_http.h>
 
-#if (T_LIMIT_REQ)
-typedef struct {
-    ngx_int_t                    index;
-    ngx_str_t                    var;
-} ngx_http_limit_req_variable_t;
-#endif
 
 typedef struct {
     u_char                       color;
@@ -43,9 +37,6 @@ typedef struct {
     ngx_uint_t                   rate;
     ngx_http_complex_value_t     key;
     ngx_http_limit_req_node_t   *node;
-#if (T_LIMIT_REQ)
-    ngx_array_t                  *limit_vars;
-#endif
 } ngx_http_limit_req_ctx_t;
 
 
@@ -54,20 +45,10 @@ typedef struct {
     /* integer value, 1 corresponds to 0.001 r/s */
     ngx_uint_t                   burst;
     ngx_uint_t                   nodelay; /* unsigned  nodelay:1 */
-#if (T_LIMIT_REQ)
-    ngx_str_t                    forbid_action;
-#endif
 } ngx_http_limit_req_limit_t;
 
 
 typedef struct {
-#if (T_LIMIT_REQ)
-    ngx_flag_t                   enable;
-
-    ngx_str_t                    geo_var_name;
-    ngx_int_t                    geo_var_index;
-    ngx_str_t                    geo_var_value;
-#endif
     ngx_array_t                  limits;
     ngx_uint_t                   limit_log_level;
     ngx_uint_t                   delay_log_level;
@@ -90,10 +71,6 @@ static char *ngx_http_limit_req_zone(ngx
     void *conf);
 static char *ngx_http_limit_req(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
-#if (T_LIMIT_REQ)
-static char *ngx_http_limit_req_whitelist(ngx_conf_t *cf, ngx_command_t *cmd,
-    void *conf);
-#endif
 static ngx_int_t ngx_http_limit_req_init(ngx_conf_t *cf);
 
 
@@ -114,9 +91,7 @@ static ngx_conf_num_bounds_t  ngx_http_l
 static ngx_command_t  ngx_http_limit_req_commands[] = {
 
     { ngx_string("limit_req_zone"),
-#if (T_LIMIT_REQ)
-      NGX_HTTP_MAIN_CONF|NGX_CONF_2MORE,
-
+      NGX_HTTP_MAIN_CONF|NGX_CONF_TAKE3,
       ngx_http_limit_req_zone,
       0,
       0,
@@ -143,15 +118,6 @@ static ngx_command_t  ngx_http_limit_req
       offsetof(ngx_http_limit_req_conf_t, status_code),
       &ngx_http_limit_req_status_bounds },
 
-#if (T_LIMIT_REQ)
-    { ngx_string("limit_req_whitelist"),
-      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE2,
-      ngx_http_limit_req_whitelist,
-      NGX_HTTP_LOC_CONF_OFFSET,
-      0,
-      NULL },
-#endif
-
       ngx_null_command
 };
 
@@ -187,93 +153,9 @@ ngx_module_t  ngx_http_limit_req_module 
 };
 
 
-#if (T_LIMIT_REQ)
-static inline ngx_int_t
-ngx_http_limit_req_ip_filter(ngx_http_request_t *r,
-    ngx_http_limit_req_conf_t *lrcf)
-{
-    ngx_http_variable_value_t    *vv;
-
-    if (lrcf->geo_var_index != NGX_CONF_UNSET) {
-        vv = ngx_http_get_indexed_variable(r, lrcf->geo_var_index);
-
-        if (vv == NULL || vv->not_found) {
-            return NGX_DECLINED;
-        }
-
-        if ((vv->len == lrcf->geo_var_value.len)
-             && (ngx_memcmp(vv->data, lrcf->geo_var_value.data, vv->len) == 0))
-        {
-            return NGX_OK;
-        }
-    }
-
-    return NGX_DECLINED;
-}
-
-
-static ngx_int_t
-ngx_http_limit_req_copy_variables(ngx_http_request_t *r, uint32_t *hash,
-    ngx_http_limit_req_ctx_t *ctx, ngx_http_limit_req_node_t *node)
-{
-    u_char                        *p;
-    size_t                         len, total_len;
-    ngx_uint_t                     i;
-    ngx_http_variable_value_t     *vv;
-    ngx_http_limit_req_variable_t *lrv;
-
-    total_len = 0;
-    p = NULL;
-
-    if (node != NULL) {
-        p = node->data;
-    }
-
-    lrv = ctx->limit_vars->elts;
-    for (i = 0; i < ctx->limit_vars->nelts; i++) {
-        vv = ngx_http_get_indexed_variable(r, lrv[i].index);
-        if (vv == NULL || vv->not_found) {
-            total_len = 0;
-            break;
-        }
-
-        len = vv->len;
-
-        if (len == 0) {
-            total_len = 0;
-            break;
-        }
-
-        if (len > 65535) {
-            ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
-                          "the value of the \"%V\" variable "
-                          "is more than 65535 bytes: \"%v\"",
-                          &lrv[i].var, vv);
-            total_len = 0;
-            break;
-        }
-
-        if (node == NULL) {
-            total_len += len;
-            ngx_crc32_update(hash, vv->data, len);
-
-        } else {
-
-            p = ngx_cpymem(p, vv->data, len);
-        }
-    }
-
-    return total_len;
-}
-#endif
-
-
 static ngx_int_t
 ngx_http_limit_req_handler(ngx_http_request_t *r)
 {
-#if (T_LIMIT_REQ)
-    size_t                       total_len;
-#endif
     uint32_t                     hash;
     ngx_str_t                    key;
     ngx_int_t                    rc;
@@ -290,17 +172,6 @@ ngx_http_limit_req_handler(ngx_http_requ
     lrcf = ngx_http_get_module_loc_conf(r, ngx_http_limit_req_module);
     limits = lrcf->limits.elts;
 
-#if (T_LIMIT_REQ)
-    if (!lrcf->enable) {
-        return NGX_DECLINED;
-    }
-
-    /* filter whitelist */
-    if (ngx_http_limit_req_ip_filter(r, lrcf) == NGX_OK) {
-        return NGX_DECLINED;
-    }
-#endif
-
     excess = 0;
 
     rc = NGX_DECLINED;
@@ -315,18 +186,6 @@ ngx_http_limit_req_handler(ngx_http_requ
 
         ctx = limit->shm_zone->data;
 
-#if (T_LIMIT_REQ)
-        ngx_str_null(&key);
-
-        ngx_crc32_init(hash);
-
-        total_len = ngx_http_limit_req_copy_variables(r, &hash, ctx, NULL);
-        if (total_len == 0) {
-            continue;
-        }
-
-        ngx_crc32_final(hash);
-#else
         if (ngx_http_complex_value(r, &ctx->key, &key) != NGX_OK) {
             return NGX_HTTP_INTERNAL_SERVER_ERROR;
         }
@@ -344,7 +203,6 @@ ngx_http_limit_req_handler(ngx_http_requ
         }
 
         hash = ngx_crc32_short(key.data, key.len);
-#endif
 
         ngx_shmtx_lock(&ctx->shpool->mutex);
 
@@ -393,32 +251,7 @@ ngx_http_limit_req_handler(ngx_http_requ
             ctx->node = NULL;
         }
 
-#if (T_LIMIT_REQ)
-        if (rc == NGX_ERROR || limit->forbid_action.len == 0) {
-            return lrcf->status_code;
-
-        } else if (limit->forbid_action.data[0] == '@') {
-
-            ngx_log_error(lrcf->limit_log_level, r->connection->log, 0,
-                          "limiting requests, forbid_action is %V",
-                          &limit->forbid_action);
-            (void) ngx_http_named_location(r, &limit->forbid_action);
-
-        } else {
-
-            ngx_log_error(lrcf->limit_log_level, r->connection->log, 0,
-                          "limiting requests, forbid_action is %V",
-                          &limit->forbid_action);
-            (void) ngx_http_internal_redirect(r,
-                                              &limit->forbid_action,
-                                              &r->args);
-        }
-
-        ngx_http_finalize_request(r, NGX_DONE);
-        return NGX_DONE;
-#else
         return lrcf->status_code;
-#endif
     }
 
     /* rc == NGX_AGAIN || rc == NGX_OK */
@@ -801,10 +634,6 @@ ngx_http_limit_req_init_zone(ngx_shm_zon
 
     size_t                     len;
     ngx_http_limit_req_ctx_t  *ctx;
-#if (T_LIMIT_REQ)
-    ngx_uint_t                     i, j;
-    ngx_http_limit_req_variable_t *v1, *v2;
-#endif
 
     ctx = shm_zone->data;
 
@@ -822,32 +651,6 @@ ngx_http_limit_req_init_zone(ngx_shm_zon
             return NGX_ERROR;
         }
 
-#if (T_LIMIT_REQ)
-        v1 = ctx->limit_vars->elts;
-        v2 = octx->limit_vars->elts;
-
-        if (ctx->limit_vars->nelts != octx->limit_vars->nelts) {
-            ngx_log_error(NGX_LOG_EMERG, shm_zone->shm.log, 0,
-                          "limit_req \"%V\" uses the \"%V\" variable "
-                          "while previously it used the \"%V\" variable",
-                          &shm_zone->shm.name, &v1[0].var, &v2[0].var);
-            return NGX_ERROR;
-        }
-
-        for (i = 0, j = 0;
-             i < ctx->limit_vars->nelts && j < octx->limit_vars->nelts;
-             i++, j++)
-        {
-            if (ngx_strcmp(v1[i].var.data, v2[j].var.data) != 0) {
-                ngx_log_error(NGX_LOG_EMERG, shm_zone->shm.log, 0,
-                              "limit_req \"%V\" uses the \"%V\" variable "
-                              "while previously it used the \"%V\" variable",
-                              &shm_zone->shm.name, &v1[i].var, &v2[j].var);
-                return NGX_ERROR;
-            }
-        }
-#endif
-
         ctx->sh = octx->sh;
         ctx->shpool = octx->shpool;
 
@@ -909,11 +712,6 @@ ngx_http_limit_req_create_conf(ngx_conf_
     conf->limit_log_level = NGX_CONF_UNSET_UINT;
     conf->status_code = NGX_CONF_UNSET_UINT;
 
-#if (T_LIMIT_REQ)
-    conf->enable = NGX_CONF_UNSET;
-    conf->geo_var_index = NGX_CONF_UNSET;
-#endif
-
     return conf;
 }
 
@@ -937,15 +735,6 @@ ngx_http_limit_req_merge_conf(ngx_conf_t
     ngx_conf_merge_uint_value(conf->status_code, prev->status_code,
                               NGX_HTTP_SERVICE_UNAVAILABLE);
 
-#if (T_LIMIT_REQ)
-    ngx_conf_merge_value(conf->enable, prev->enable, 0);
-
-    ngx_conf_merge_value(conf->geo_var_index, prev->geo_var_index,
-                         NGX_CONF_UNSET);
-
-    ngx_conf_merge_str_value(conf->geo_var_value, prev->geo_var_value, "");
-#endif
-
     return NGX_CONF_OK;
 }
 
@@ -961,20 +750,7 @@ ngx_http_limit_req_zone(ngx_conf_t *cf, 
     ngx_uint_t                         i;
     ngx_shm_zone_t                    *shm_zone;
     ngx_http_limit_req_ctx_t          *ctx;
-#if (T_LIMIT_REQ)
-    ngx_array_t                       *variables;
-    ngx_http_limit_req_variable_t     *v;
-#else
     ngx_http_compile_complex_value_t   ccv;
-#endif
-
-#if (T_LIMIT_REQ)
-    variables = ngx_array_create(cf->pool, 5,
-                                 sizeof(ngx_http_limit_req_variable_t));
-    if (variables == NULL) {
-        return NGX_CONF_ERROR;
-    }
-#endif
 
     value = cf->args->elts;
 
@@ -983,23 +759,6 @@ ngx_http_limit_req_zone(ngx_conf_t *cf, 
         return NGX_CONF_ERROR;
     }
 
-#if (T_LIMIT_REQ)
-    v = ngx_array_push(variables);
-    if (v == NULL) {
-        return NGX_CONF_ERROR;
-    }
-
-    value[1].len--;
-    value[1].data++;
-
-    v->index = ngx_http_get_variable_index(cf, &value[1]);
-    if (v->index == NGX_ERROR) {
-        return NGX_CONF_ERROR;
-    }
-
-    v->var = value[1];
-
-#else
     ngx_memzero(&ccv, sizeof(ngx_http_compile_complex_value_t));
 
     ccv.cf = cf;
@@ -1009,7 +768,6 @@ ngx_http_limit_req_zone(ngx_conf_t *cf, 
     if (ngx_http_compile_complex_value(&ccv) != NGX_OK) {
         return NGX_CONF_ERROR;
     }
-#endif
 
     size = 0;
     rate = 1;
@@ -1076,28 +834,6 @@ ngx_http_limit_req_zone(ngx_conf_t *cf, 
             continue;
         }
 
-#if (T_LIMIT_REQ)
-        if (value[i].data[0] == '$') {
-
-            value[i].len--;
-            value[i].data++;
-
-            v = ngx_array_push(variables);
-            if (v == NULL) {
-                return NGX_CONF_ERROR;
-            }
-
-            v->index = ngx_http_get_variable_index(cf, &value[i]);
-            if (v->index == NGX_ERROR) {
-                return NGX_CONF_ERROR;
-            }
-
-            v->var = value[i];
-
-            continue;
-        }
-#endif
-
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
                            "invalid parameter \"%V\"", &value[i]);
         return NGX_CONF_ERROR;
@@ -1112,10 +848,6 @@ ngx_http_limit_req_zone(ngx_conf_t *cf, 
 
     ctx->rate = rate * 1000 / scale;
 
-#if (T_LIMIT_REQ)
-    ctx->limit_vars = variables;
-#endif
-
     shm_zone = ngx_shared_memory_add(cf, &name, size,
                                      &ngx_http_limit_req_module);
     if (shm_zone == NULL) {
@@ -1148,24 +880,9 @@ ngx_http_limit_req(ngx_conf_t *cf, ngx_c
     ngx_uint_t                   i, nodelay;
     ngx_shm_zone_t              *shm_zone;
     ngx_http_limit_req_limit_t  *limit, *limits;
-#if (T_LIMIT_REQ)
-    ngx_str_t                    forbid_action;
-#endif
 
     value = cf->args->elts;
 
-#if (T_LIMIT_REQ)
-    if (cf->args->nelts == 2) {
-        if (ngx_strncmp(value[1].data, "off", 3) == 0) {
-            lrcf->enable = 0;
-            return NGX_CONF_OK;
-        }
-    }
-
-    lrcf->enable = 1;
-    ngx_str_null(&forbid_action);
-#endif
-
     shm_zone = NULL;
     burst = 0;
     nodelay = 0;
@@ -1203,24 +920,6 @@ ngx_http_limit_req(ngx_conf_t *cf, ngx_c
             continue;
         }
 
-#if (T_LIMIT_REQ)
-        if (ngx_strncmp(value[i].data, "forbid_action=", 14) == 0) {
-
-            s.len = value[i].len - 14;
-            s.data = value[i].data + 14;
-
-            if (s.len < 2 || (s.data[0] != '@' && s.data[0] != '/')) {
-                ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                                   "invalid forbid_action \"%V\"", &value[i]);
-                return NGX_CONF_ERROR;
-            }
-
-            forbid_action = s;
-
-            continue;
-        }
-#endif
-
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
                            "invalid parameter \"%V\"", &value[i]);
         return NGX_CONF_ERROR;
@@ -1258,62 +957,9 @@ ngx_http_limit_req(ngx_conf_t *cf, ngx_c
     limit->shm_zone = shm_zone;
     limit->burst = burst * 1000;
     limit->nodelay = nodelay;
-#if (T_LIMIT_REQ)
-    limit->forbid_action = forbid_action;
-#endif
-
-    return NGX_CONF_OK;
-}
-
-
-#if (T_LIMIT_REQ)
-static char *
-ngx_http_limit_req_whitelist(ngx_conf_t *cf, ngx_command_t *cmd,
-    void *conf)
-{
-    ngx_http_limit_req_conf_t  *lrcf = conf;
-
-    ngx_str_t              *value, s;
-    ngx_uint_t              i;
-
-    value = cf->args->elts;
-
-    for (i = 1; i < cf->args->nelts; i++) {
-        if (ngx_strncmp(value[i].data, "geo_var_name=", 13) == 0) {
-
-            s.len = value[i].len - 13;
-            s.data = value[i].data + 13;
-
-            lrcf->geo_var_name = s;
-
-            lrcf->geo_var_index = ngx_http_get_variable_index(cf,
-                &lrcf->geo_var_name);
-
-            if (lrcf->geo_var_index == NGX_ERROR) {
-                return NGX_CONF_ERROR;
-            }
-
-            continue;
-        }
-
-        if (ngx_strncmp(value[i].data, "geo_var_value=", 14) == 0) {
-
-            s.len = value[i].len - 14;
-            s.data = value[i].data + 14;
-
-            lrcf->geo_var_value = s;
-
-            continue;
-        }
-
-        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                           "invalid parameter \"%V\"", &value[i]);
-        return NGX_CONF_ERROR;
-    }
 
     return NGX_CONF_OK;
 }
-#endif
 
 
 static ngx_int_t

```

**Pull request closed without merging**